SOLR-12697 Add pure DocValues support to FieldValueFeature

tomglk opened the pull request:
<!--
_(If you are a project committer then you may remove some/all of the following template.)_

Before creating a pull request, please file an issue in the ASF Jira system for Solr:

* https://issues.apache.org/jira/projects/SOLR

You will need to create an account in Jira in order to create an issue.

The title of the PR should reference the Jira issue number in the form:

* SOLR-####: <short description of problem or changes>

SOLR must be fully capitalized. A short description helps people scanning pull requests for items they can work on.

Properly referencing the issue in the title ensures that Jira is correctly updated with code review comments and commits. -->


# Description

This PR refers to the issue [SOLR-12697](https://issues.apache.org/jira/browse/SOLR-12697).
The problem is, that the current FieldValueFeature only works for stored fields.
This is not optimal, because using DocValues is faster for this use case. Also it increases the index size if you have to store fields only to use them for ltr.

# Solution

**Note:** This PR is based on the work of Stanislav Livotov and Christine Poerschke that can be seen in the jira ticket.

It uses the latest patch (17th May 2019) from the jira ticket as base. I combined that with suggestions from Mrs. Poerschke and my own approach to the problem.

This PR adds the DocValuesFieldValueFeatureScorer as a new Scorer used by the FieldValueFeatureWeight.
The new scorer is used whenever a field has docValues and is not stored. Therefore it does not affect the current functionality but only is applied for fields that could not be used before.
The new scorer checks the type of docValues a field has and handles NUMERIC and SORTED types. For NUMERIC fields, it simply uses the value, the SORTED type gets parsed as number or boolean-flag.

# Tests

New fields that have docValues=true were added to the schema.xml in order to test in TestLTROnSolrCloud that the feature-requests also return values for these fields.
The TestLTRReRankingPipeline was changed from a SolrTestCase  to a SolrTestCaseJ4 in order to improve readability.

I ran all tests in the package `org.apache.solr.ltr`.

# Please note

I am aware that the structure of the FieldValueFeature is now quite hard to read and the new Scorer is a bit hidden. I decided to add another nested class to the FieldValueFeatureWeight to avoid having to duplicate a lot of code just to change the inner functionality.

~~Unit tests for the handleBytesRef are still missing. I plan to add them, but wanted to create the PR already so that the general approach can be reviewed and discussed.~~
Edit: Tests for the parsing of different Types of sorted docValues (T and F flag, numbers, no number) were added in e07c432.

# Checklist

Please review the following and check all that apply:

- [X] I have reviewed the guidelines for [How to Contribute](https://wiki.apache.org/solr/HowToContribute) and my code conforms to the standards described there to the best of my ability.
- [X] I have created a Jira issue and added the issue ID to my pull request title. (**Issue was already present**)
- [X] I have given Solr maintainers [access](https://help.github.com/en/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork) to contribute to my PR branch. (optional but recommended)
- [X] I have developed this patch against the `main` branch.
- [ ] I have run `./gradlew check`.
- [X] I have run `./gradlew check -x test`.
- [X] I have added tests for my changes.
- [ ] I have added documentation for the [Reference Guide](https://github.com/apache/solr/tree/main/solr/solr-ref-guide)

cpoerschke commented on solr/contrib/ltr/src/test/org/apache/solr/ltr/TestLTROnSolrCloud.java:
```diff
@@ -232,32 +230,47 @@ private void createCollection(String name, String config, int numShards, int num
+    for (int docId = collectionSize; docId >= 1;  docId--) {
       final int popularity = docId;
       indexDocument(collection, String.valueOf(docId), "a1", "bloom", popularity);
+      if(docId == collectionSize / 2) {
+        // commit in the middle in order to check that everything works fine for multi-segment case
+        solrCluster.getSolrClient().commit(collection);
+      }
```
It's great to expand the test coverage here to check that the multi-segment case works. However _always_ having the commit in the middle also changes what the existing test does i.e. the single-segment or fewer segments scenario. A _"best of both worlds"_ way could be use of randomisation i.e. only _sometimes_ commit in the middle e.g.

```
  if (random.nextBoolean()) solrCluster.getSolrClient().commit(collection);
```

which is supported by the test framework.

tomglk replied:
Thanks for the suggestion with nextBoolean(), I didn't know that. Used it in 4559415.

cpoerschke commented on solr/contrib/ltr/src/test/org/apache/solr/ltr/TestLTROnSolrCloud.java:
```diff
@@ -232,32 +230,47 @@ private void createCollection(String name, String config, int numShards, int num
+      doc.setField("dvStrBoolField", popularity % 2 == 0 ? "T" : "F");
+    }
     solrCluster.getSolrClient().add(collection, doc);
   }
 
   private void indexDocuments(final String collection)
        throws Exception {
     final int collectionSize = 8;
-    for (int docId = 1; docId <= collectionSize;  docId++) {
+    // put documents in reversed order to check that advanceExact is working correctly
+    for (int docId = collectionSize; docId >= 1;  docId--) {
```
Similar to multi-segment case coverage below, randomisation could be used here to cover both ascending and descending order. Or another way could be to form a list of ids and to shuffle it with the randomiser and then to use it to index.

tomglk replied:
I like your idea of shuffling the ids and used it in 4559415.

cpoerschke reviewed (commented):
Thanks @tomglk for opening this pull request today!

I've added some comments from an initial review and hope to look more fully maybe later this week.

@alessandrobenedetti might also be interested in this.

tomglk commented:
Thank you very much for the quick and detailed feedback, @cpoerschke !
I already have some ideas in mind and will start to improve the code next thing tomorrow morning.

tomglk commented on solr/contrib/ltr/src/java/org/apache/solr/ltr/feature/FieldValueFeature.java:
```diff
@@ -146,5 +171,100 @@ public float getMaxScore(int upTo) throws IOException {
+            case SORTED:
+              docValues = DocValues.getSorted(context.reader(), field);
+              break;
+            case BINARY:
+            case SORTED_NUMERIC:
+            case SORTED_SET:
+            case NONE:
+            default:
+              docValues = null;
```
I tried using `schemaField.docValuesType()`, but it always returned NONE. I am not sure why.
That's why I used the fieldInfos to get the Type.

tomglk replied:
The refactoring in ec4cbfb moves the determination of the docValuesType prior to the construction of the `DocValuesFieldValueFeatureScorer`.
This way, `score()` only has to handle the two supported types and the long switch-cases could be removed.
`docValues == null` would produce a NPE, but that case cannot happen.

tomglk commented on solr/contrib/ltr/src/java/org/apache/solr/ltr/feature/FieldValueFeature.java:
```diff
@@ -146,5 +171,100 @@ public float getMaxScore(int upTo) throws IOException {
+              if (NumberType.FLOAT.equals(schemaFieldType.getNumberType())) {
+                // convert float value that was stored as long back to float
+                return Float.intBitsToFloat((int) ((NumericDocValues) docValues).longValue());
+              } else if (NumberType.DOUBLE.equals(schemaFieldType.getNumberType())) {
```
Valid point. I applied your suggestion in commit 5bc995c.

tomglk commented on solr/contrib/ltr/src/java/org/apache/solr/ltr/feature/FieldValueFeature.java:
```diff
@@ -146,5 +171,100 @@ public float getMaxScore(int upTo) throws IOException {
+      DocValuesType docValuesType = DocValuesType.NONE;
+
+      public DocValuesFieldValueFeatureScorer(final FeatureWeight weight, final LeafReaderContext context,
+                                              final DocIdSetIterator itr, final SchemaField schemaField) {
```
Sounds good to me, changed it in 5bc995c.

tomglk commented on solr/contrib/ltr/src/test-files/solr/collection1/conf/schema.xml:
```diff
@@ -18,13 +18,22 @@
-    <field name="id" type="string" indexed="true" stored="true" required="true" multiValued="false" />
+    <field name="id" type="string" indexed="true" stored="true" required="true" multiValued="false"/>
+    <field name="final-score" type="string" indexed="true" stored="true" multiValued="false"/>
+    <field name="final-score-float" type="float" indexed="true" stored="true" multiValued="false"/>
```
Changed it to camel case in 4559415.

tomglk commented on solr/contrib/ltr/src/test/org/apache/solr/ltr/TestLTROnSolrCloud.java:
```diff
@@ -107,21 +105,21 @@ public void testSimpleQuery() throws Exception {
+        "powpularityS", "4.0", "c3","2.0", "original","6.0", "dvIntFieldFeature","2.0","dvLongFieldFeature","2.0","dvFloatFieldFeature","0.2","dvDoubleFieldFeature","0.2","dvStrNumFieldFeature","2.0","dvStrBoolFieldFeature","1.0");
     final String result7_features= FeatureLoggerTestUtils.toFeatureVector(
-        "powpularityS", "1.0", "c3","2.0", "original","7.0");
+        "powpularityS", "1.0", "c3","2.0", "original","7.0", "dvIntFieldFeature","-1.0","dvLongFieldFeature","-2.0","dvFloatFieldFeature","-3.0","dvDoubleFieldFeature","-4.0","dvStrNumFieldFeature","-5.0","dvStrBoolFieldFeature","0.0");
```
Sure thing! Wrapped the lines in 4559415.

tomglk commented on solr/contrib/ltr/src/java/org/apache/solr/ltr/feature/FieldValueFeature.java:
```diff
@@ -146,5 +171,100 @@ public float getMaxScore(int upTo) throws IOException {
+     */
+    public class DocValuesFieldValueFeatureScorer extends FeatureWeight.FeatureScorer {
+      final LeafReaderContext context;
+      final DocIdSetIterator docValues;
```
Sadly, the `DocValuesIterator` is not public, so I cannot use it here.
But I am working on a restructuring.

tomglk replied:
The restructuring happened in ec4cbfb.
I am not quite sure, whether I like the duplication in lines 216 & 219, but by reversing the check for the type and the check for the presence of the docId, the cast to the correct docValuesType can happen earlier, allowing the use of `advanceExact()`.

`advanceExact()` generally seems to be preferred over `advance()`, can you tell me why?

cpoerschke replied:
> ... `advanceExact()` generally seems to be preferred over `advance()`, can you tell me why?

Good question, I don't know!

Intuitively something that returns `boolean` rather than `int` when the `int` is not needed seems neater but really we are after performance and not necessarily neatness.

`NumericDocValues.longValue()` mentioned `advanceExact()` so that could be a clue but not an explanation.

Slightly random code browsing to (say) https://github.com/apache/lucene-solr/blob/releases/lucene-solr/8.8.2/lucene/core/src/java/org/apache/lucene/codecs/lucene80/Lucene80DocValuesProducer.java finds some delegation to https://github.com/apache/lucene-solr/blob/releases/lucene-solr/8.8.2/lucene/core/src/java/org/apache/lucene/codecs/lucene80/IndexedDISI.java and its advanceNext method and mentions of `IndexedDISI` always make me a bit _dizzy_ since I don't yet know much about it.

The last commit on https://github.com/apache/lucene-solr/blob/releases/lucene-solr/8.8.2/lucene/core/src/java/org/apache/lucene/index/NumericDocValues.java reads "LUCENE-7462: Give doc values APIs an advanceExact method." and that commit and https://issues.apache.org/jira/browse/LUCENE-7462 should have more info then.

tomglk replied:
Thank you for the links. :)
I only had some time to skim trough it, but this really looks interesting. As soon as my time allows it, I will definitely dig deeper into that.

cpoerschke commented on solr/contrib/ltr/src/test/org/apache/solr/ltr/TestLTROnSolrCloud.java:
```diff
@@ -232,58 +249,113 @@ private void createCollection(String name, String config, int numShards, int num
     final int collectionSize = 8;
-    for (int docId = 1; docId <= collectionSize;  docId++) {
+    // put documents in random order to check that advanceExact is working correctly
+    List<Integer> docIds = IntStream.rangeClosed(1, collectionSize).boxed().collect(toList());
```
Nice, I haven't across this `IntStream.rangeClosed` thing before!

cpoerschke commented on solr/contrib/ltr/src/test/org/apache/solr/ltr/TestLTROnSolrCloud.java:
```diff
@@ -232,58 +249,113 @@ private void createCollection(String name, String config, int numShards, int num
     doc.setField("title", title);
     doc.setField("description", description);
     doc.setField("popularity", popularity);
+    if (popularity != 1) {
+      // check that empty values will be read as default
+      doc.setField("dvIntField", popularity);
+      doc.setField("dvLongField", popularity);
+      doc.setField("dvFloatField", ((float) popularity) / 10);
+      doc.setField("dvDoubleField", ((double) popularity) / 10);
+      doc.setField("dvStrNumField", popularity);
+      doc.setField("dvStrBoolField", popularity % 2 == 0 ? "T" : "F");
+    }
```
I looked at `TestLTROnSolrCloud` changes today, this approach here is very elegant, combined with the `resultN_features` and `schema.xml` change.

cpoerschke commented on solr/contrib/ltr/src/test-files/solr/collection1/conf/schema.xml:
```diff
@@ -18,13 +18,22 @@
     <field name="title" type="text_general" indexed="true" stored="true"/>
     <field name="description" type="text_general" indexed="true" stored="true"/>
     <field name="keywords" type="text_general" indexed="true" stored="true" multiValued="true"/>
     <field name="popularity" type="int" indexed="true" stored="true" />
```
observation: via the `docValues="${solr.tests.numeric.dv}"` in the `<fieldType name="int" ...` and `popularity` being used in the `TestFieldValueFeature` we have some existing test coverage for doc values.

cpoerschke replied:
Added a commit with additional `TestFieldValueFeature` coverage, for the numeric fields only so far i.e. `dvStrNumPopularity` and `dvStrBoolPopularity` not yet.

cpoerschke commented on solr/contrib/ltr/src/java/org/apache/solr/ltr/feature/FieldValueFeature.java:
```diff
@@ -57,50 +67,79 @@ public void setField(String field) {
+        FieldInfo fieldInfo = context.reader().getFieldInfos().fieldInfo(field);
+        DocValuesType docValuesType = fieldInfo != null ? fieldInfo.getDocValuesType() : DocValuesType.NONE;
+
+        if (DocValuesType.NUMERIC.equals(docValuesType)) {
+          return new NumericDocValuesFieldValueFeatureScorer(this, context,
+                  DocIdSetIterator.all(DocIdSetIterator.NO_MORE_DOCS), schemaField.getType().getNumberType());
+        } else if (DocValuesType.SORTED.equals(docValuesType)) {
+          return new SortedDocValuesFieldValueFeatureScorer(this, context,
+                  DocIdSetIterator.all(DocIdSetIterator.NO_MORE_DOCS));
+          // If type is NONE, this segment has no docs with this field. That's not a problem, because we won't call score() anyway
```
Have added a commit to split the dual purpose `DocValuesFieldValueFeatureScorer` into two, I think it helps reduce type casting and this-or-that choices (and the class members required to make those choices) in the DocValuesFieldValueFeatureScorer implementation, what do you think?

tomglk replied:
I think that's a great change, it really improves the readability!

cpoerschke commented on solr/contrib/ltr/src/test-files/solr/collection1/conf/schema.xml:
```diff
@@ -38,6 +51,11 @@
   <copyField source="title" dest="text"/>
   <copyField source="description" dest="text"/>
 
+  <copyField source="popularity" dest="dvIntPopularity"/>
+  <copyField source="popularity" dest="dvLongPopularity"/>
+  <copyField source="popularity" dest="dvFloatPopularity"/>
+  <copyField source="popularity" dest="dvDoublePopularity"/>
```
observation: `copyField` here works for the `dv*Popularity` fields but the `dv(Int|Long|Float|Double)Field` fields cannot use it since the defaulting logic and the popularity/10 logic is happening there.

cpoerschke commented on solr/contrib/ltr/src/test/org/apache/solr/ltr/feature/TestFieldValueFeature.java:
```diff
@@ -70,86 +78,92 @@ public void after() throws Exception {
+      query.add("rq", "{!ltr model="+field+"-model reRankDocs=4}");
+      query.add("fl", "[fv]");
+      assertJQ("/query" + query.toQueryString(), "/response/numFound/==1");
+      assertJQ("/query" + query.toQueryString(),
+              "/response/docs/[0]/=={'[fv]':'"+FeatureLoggerTestUtils.toFeatureVector(field,Float.toString(FIELD_VALUE_FEATURE_DEFAULT_VAL))+"'}");
```
**This test fails here on the second and subsequent run of the loop.** I don't know yet if that is a test artifact or an actual code issue. The `testIfADocumentDoesntHaveAFieldASetDefaultValueIsReturned` test does not have the same behaviour since each loop turn uses a different `f(feature )store` (and its own model) whereas this test uses the test-wide models and its default feature store.

tomglk replied:
I think the problem is `final private boolean extractAllFeatures;` in `LTRScoringQuery`. It is set to true and therefore all previous features are added to the response instead of just the model features.

This is the logic (line 210 in LTRScoringQuery):
```java
if (this.extractAllFeatures) {
    features = allFeatures;
} else{
    features =  modelFeatures;
}
```

I didn't find a way to set it to false with a query param.
So we would have two options:
- use this constructor `public LTRScoringQuery(LTRScoringModel ltrScoringModel, boolean extractAllFeatures)`
- adjust the xpath to something like `[ends-with(., '<fieldname>=0.0')]`

I am trying to adjust the xpath, because I would prefer that solution, but I was not able to find the right pattern so far.

cpoerschke replied:
1/2 Thanks for investigating in detail here! I agree the explanation is with the `extractAllFeatures` boolean i.e. the response contains the values for all the features in the feature store. The puzzling thing then though is why the first run of the loop succeeds and only subsequent runs fail. Since all loop runs use the same feature store and surely each loop run should return the same feature values? ...

cpoerschke reviewed (commented):
Hi @tomglk. Thanks for working on this pull request! I've added two commits to the branch, let me know what you think? Yet to look at the `TestLTRReRankingPipeline` test (next week).

cpoerschke commented on solr/contrib/ltr/src/java/org/apache/solr/ltr/feature/FieldValueFeature.java:
```diff
@@ -106,18 +106,29 @@ public FieldValueFeatureWeight(IndexSearcher searcher,
+        if (DocValuesType.NUMERIC.equals(docValuesType) || DocValuesType.SORTED.equals(docValuesType)) {
+          return new DocValuesFieldValueFeatureScorer(this, context,
+                  DocIdSetIterator.all(DocIdSetIterator.NO_MORE_DOCS), schemaField.getType(), docValuesType);
+          // If type is NONE, this segment has no docs with this field. That's not a problem, because we won't call score() anyway
```
Earlier today I had a thought on this code area here and wanted to note it down (for further pondering next week) but had to repair my bicycle's puncture first. Anyway ...

The _"If type is NONE, this segment has no docs with this field. That's not a problem ..."_ comment here is very helpful. I wonder though if the _"because we won't call score() anyway"_ part remains accurate after the refactorings. In the branch's initial commit I think it was accurate i.e. via the `if (docValues != null` at the start of the `score()` implementation the inside of the `if` block wasn't called. But now there's something more interesting going on here:

* If there's a segment whose documents all don't have the field then instead of a `DocValuesFieldValueFeatureScorer` the `FieldValueFeatureScorer` is used via the fall-through.
* Conceptually that jumps out as untidy or unintuitive i.e. if there is no field for all the documents in the segment then (say) a `NoFieldValueFeatureScorer` is more obvious.
* In terms of performance use of the `FieldValueFeatureScorer` should be more costly i.e. reading of the field would be attempted and nothing would be found.
* In terms of test coverage this would not be caught since the default value is returned, albeit inefficiently, by the `FieldValueFeatureScorer` implementation.
* In terms of use cases this code path may or may not be an edge case e.g. if a `FieldValueFeature` is for a field that is sparsely populated then a good proportion of segments might encounter the code path and reasons for sparseness may include (say) historical documents not having the field.

_If_ that reasoning makes sense then a next step here could be to create and use a `NoFieldValueFeatureScorer` class. What do you think?

tomglk replied:
Great observation! I completely overlooked this.
Your reasoning makes complete sense to me and some debugging verified the behavior you described.

Even if it may be an edge case, I think it makes sense to make it explicit.
In terms of performance you are also right. Trying to read the stored fields before returning the default value surely is more costly than just returning the value.

I started to create a `DefaultValueFieldValueFeatureScorer` that always returns the default value, but until now I am struggling to find a way to test this.

I hope your bike is well now! :)

cpoerschke replied:
> ... to find a way to test this. ...

Thinking conceptually (without actually trying it) one way could be like this ...
* imagine `TestFieldValueFeature` had a static inner class `FooBarFieldValueFeature extends FieldValueFeature` and for some of its tests it uses that instead of the 'real' `FieldValueFeature` class
* imagine `FooBarFieldValueFeature` can observe whatever needs to be observed and capture the observations in class level variables
* if the class level variables are public or package scope the test has access to them

Well, that's the theory but in practice what we wish to observe and test is not directly in `FieldValueFeature` but inside the `FieldValueFeatureWeight` created, but still it could be possible e.g. something like

```
class FooBarFieldValueFeature extends FieldValueFeature {

  static AtomicBoolean seenFieldValueFeatureScorer = false; // could also be counter instead of boolean

  @Override
  public FeatureWeight createWeight(...) throws IOException {
    return new FooBarFieldValueFeatureWeight(...);
  }

  public class FooBarFieldValueFeatureWeight extends FieldValueFeatureWeight {
    @Override
    public FeatureScorer scorer(LeafReaderContext context) throws IOException {
      FeatureScorer featureScorer = super.scorer();
      if (featureScorer instanceof FieldValueFeatureScorer) {
        seenFieldValueFeatureScorer = true;
      }
      return featureScorer;
    }
  }
}
```

structurally and then in the test `assertFalse(FooBarFieldValueFeature.seenFieldValueFeatureScorer);`

Does that kind of convey the idea? Essentially the idea is to extend the "real" class with a "test" class, for the test class to delegate to the real class and for the test class to be "shaped" to suit the test's purposes.

tomglk replied:
That's a great idea!
I wanted to include the test in the `TestLTROnSolrCloud`, but then I needed to control which documents are in which segment to be able to (not) trigger the fallback. That didn't work.

tomglk replied:
This was implemented in da6a635 with the `ObservingFieldValueFeature`.

cpoerschke commented on solr/contrib/ltr/src/test/org/apache/solr/ltr/feature/TestFieldValueFeature.java:
```diff
@@ -56,11 +62,13 @@ public void before() throws Exception {
-            "{\"field\":\"popularity\"}");
+    for (String field : FIELD_NAMES) {
+      loadFeature(field, FieldValueFeature.class.getName(),
+              "{\"field\":\""+field+"\"}");
 
-    loadModel("popularity-model", LinearModel.class.getName(),
-            new String[] {"popularity"}, "{\"weights\":{\"popularity\":1.0}}");
+      loadModel(field + "-model", LinearModel.class.getName(),
+              new String[] {field}, "{\"weights\":{\""+field+"\":1.0}}");
+    }
```
2/2 ... I think the explanation for the mystery is here i.e. the `LinearModel` constructor signature and the interleaving of feature and model loading results in the models having different definitions of what `allFeatures` means. One solution then would be for the test expectations to account for the multiple meanings of `allFeatures` or (my preference probably) for there to be only one `allFeatures` meaning e.g. via the separation of feature and model loading. Or perhaps you can think of a third solution even?

```
for (String field : FIELD_NAMES) {
  loadFeature(field, ...);
}
for (String field : FIELD_NAMES) {
  loadModel(field + "-model", ...);
}
```

tomglk replied:
Thanks for continuing the investigation!
I think the example you posted here should work, although I am still a bit irritated regarding the `extractAllFeatures = true`.

As user I would expect it to always return all the features that were created, no matter if that happened before or after the model. 
But the code makes it explicit, that these information should not be modified (`allFeatures` is an unmodifiableList).
I think that a nicely placed docString could avoid some confusion.

cpoerschke replied:
Yes, it's a somewhat quirky behaviour. Operationally it could also cause mysteries e.g. when new features are added to an existing store and the (reasonable) expectation is that they will start producing feature values right away when actually that only happens when the existing model(s) are next re-instantiated. +1 to try and reduce such confusion via documentation somehow.

tomglk commented:
I have added more tests and an entry to the changes.txt.
For now, I just added our names, @cpoerschke , because I wasn't sure about the people in the Jira ticket. I think it would be nice to include them because this PR was built upon that and the patches, but I have no overview who participated how much and the last patch was created ~ 2 years ago.
What do you think?

Also: From my side this looks quite finished now. Do you miss anything?

cpoerschke commented on solr/contrib/ltr/src/test/org/apache/solr/ltr/feature/TestFieldValueFeature.java:
```diff
@@ -213,4 +349,69 @@ public void testParamsToMap() throws Exception {
+    }
+  }
+
+  /**
+   * This class is used to track which specific FieldValueFeature is used so that we can test, whether the
+   * fallback mechanism works correctly.
+   */
+  public static class ObservingFieldValueFeature extends FieldValueFeature {
```
The selection of test coverage using this new `ObservingFieldValueFeature` and its observing the scorer type is very comprehensive and elegant, thanks for adding it!

cpoerschke commented on solr/contrib/ltr/src/java/org/apache/solr/ltr/feature/FieldValueFeature.java:
```diff
@@ -23,24 +23,34 @@
  * <pre>{
-  "name":  "rawHits",
-  "class": "org.apache.solr.ltr.feature.FieldValueFeature",
-  "params": {
-      "field": "hits"
-  }
-}</pre>
+ * "name":  "rawHits",
+ * "class": "org.apache.solr.ltr.feature.FieldValueFeature",
+ * "params": {
+ * "field": "hits",
+ * "defaultValue": -1
+ * }
+ * }</pre>
  */
```
Doing the _"a.k.a. pure DocValues"_ edit on the `solr/CHANGES.txt` made me wonder, would it be helpful to have more detail here w.r.t. (stored and docValues) field type attributes, what do you think?

tomglk replied:
I think it is always helpful to make such branching in the code explicit.
So I added some javadoc to explain, which scorer is used for which type of field. (b9d3cd0)

cpoerschke commented on solr/contrib/ltr/src/test/org/apache/solr/ltr/TestLTRReRankingPipeline.java:
```diff
@@ -41,25 +37,33 @@
 import org.junit.Test;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
```
After multiple attempts I'm still struggling to understand and appreciate how the `TestLTRReRankingPipeline` changes relate to and fit in with the other changes. @tomglk would you have an extra perspective on this perhaps?

I'm speculating that the changes were required when the `SolrDocumentFetcher` implementation was used but now with the `TestFieldValueFeature` changes the picture has changed.

_If_ the `TestLTRReRankingPipeline` changes are no longer required and relatively unrelated I would favour `git checkout origin/main -- TestLTRReRankingPipeline.java` removing them from the pull request scope. What do you think?

cpoerschke replied:
Thanks for this! Yes, once the "where is the pure docValues connection?" question is gone and the mental focus is on "code comprehension and readability" it is clear that this is a refactoring change that is indeed very beneficial.

And when considering if the refactoring changes inadvertently changes any existing behaviour the `@AwaitsFix(bugUrl = "https://issues.apache.org/jira/browse/SOLR-11134")` caught my eye and I think I managed to diagnose this long-standing issue:
* the problem was that `new FloatDocValuesField("final-score", ...)` resulted in a `stored=false` field (which was not very apparent) and its `schema.xml` replacement named `finalScore` is `stored=true` i.e. immediate readability and comprehension benefit there.
* then next the `testDifferentTopN` turned out to have an incorrect expectation w.r.t. what the score should be (addition of an alternative `TestLinearModel.makeFeatureWeights` variant tries to make that clearer).
* and lastly once the `testDifferentTopN` test passed it interacted with the `testRescorer` test i.e. if the former ran first then the latter would see not only its own 2 documents but also the documents added by the former test, this is easily fixed by everyone deleting all documents before adding their documents.

> ... So I would like to keep it in this PR to make it more accessible and easier to built up upon in the future.

I agree we should keep these very beneficial `TestLTRReRankingPipeline` changes.

If (and that is an if) the SOLR-11134 fix (or something along those lines) makes sense, then how would you feel about us still removing the `TestLTRReRankingPipeline` change from this PR and instead to have it in a separate PR and to "book it" only under SOLR-11134 or jointly under SOLR-11134 and SOLR-12697 tickets?

tomglk replied:
Thank you for taking the time to investigate this!
I just had a look at your changes and they totally make sense to me.

Because of the existing jira ticket for 11134 it makes sense to me to remove the file from this PR and to book it under a new PR like "SOLR-11134 fix and restructure TestLTRReRankingPipeline.testDifferentTopN test".

cpoerschke replied:
Thanks for looking at and reviewing the changes! I've just opened #145 PR for the "SOLR-11134 fix and restructure TestLTRReRankingPipeline.testDifferentTopN test" changes then.

cpoerschke reviewed (commented):
Hi @tomglk!

> ... From my side this looks quite finished now. Do you miss anything?

I agree, we're close to wrapping this up (and am also mindful of the 8.9.0 release coming up soon and us wishing to include this in it). There's a couple of "polishing" bits I'd like to do and me adding small commits directly to the branch (like the three commits just added) might be the most efficient way if that works for you?

Really looking forward to this being in the next release (and also have some ideas on how afterwards we potentially might (under separate JIRA and pull request) (safely) break backwards compatibility in Solr 9.0.0 to allow `stored=true docValues=true` fields to also benefit from the docValues implementation whilst a deprecated (say) `LegacyFieldValueFeature` until (say) Solr 10.0.0 retains the existing implementation for users who need that ...

tomglk commented:
Hi @cpoerschke ,
feel free to directly commit to the branch. :)

Your idea to make a new jira ticket and PR to break backwards compatibility sounds good to me. It would be useful to profit from the better performance for more fields. I am on board to continue working on this topic with you.

I'm also exited to see this in solr 8.9.0!

tomglk commented:
I am currently debugging some ltr stuff and stumbled upon this:
``` java
final String string = indexableField.stringValue();
if (string.length() == 1) {
    // boolean values in the index are encoded with a single char contained in TRUE_TOKEN or FALSE_TOKEN
    // (see BoolField)
    if (string.charAt(0) == BoolField.TRUE_TOKEN[0]) {
      return 1f;
    }
    if (string.charAt(0) == BoolField.FALSE_TOKEN[0]) {
      return 0f;
    }
}
```
This code is in the `FieldValueFeatureScorer`. Why don't we try to parse the string value as number?
We do that for docValues, but not for stored fields.

cpoerschke commented:
Hi @tomglk!

> I am currently debugging some ltr stuff and stumbled upon this: ... Why don't we try to parse the string value as number? We do that for docValues, but not for stored fields.

Ah, we seem to have both stumbled upon the same question though perhaps via different routes?!

I was trying to extend `TestFieldValueFeature.testThatStringValuesAreCorrectlyParsed` to also cover non-docValues field _and_ to ensure behavioural consistency between dv and non-dv fields i.e. docValues and stored fields. My conclusion was that for back compat reasons the docValues would need to match stored fields behaviour -- _I will add a commit with proposed changes shortly_ -- and that what looks like a "string" value in the implementation is for "boolean" values only really i.e. F/0 and T/1 and when number values are required a number field type would/should be used instead.

_And advance apologies, the commit has multiple things in one commit though I've tried to at least describe in the commit message details what is in it. Looking forward to your thoughts on it._

cpoerschke commented on solr/contrib/ltr/src/java/org/apache/solr/ltr/feature/FieldValueFeature.java:
```diff
@@ -146,5 +201,137 @@ public float getMaxScore(int upTo) throws IOException {
+       * @return The numeric value that the docValues contain for the current document
+       * @throws IOException if docValues cannot be read
+       */
+      private float readNumericDocValues() throws IOException {
+        if (NumberType.FLOAT.equals(numberType)) {
+          // convert float value that was stored as long back to float
+          return Float.intBitsToFloat((int) docValues.longValue());
+        } else if (NumberType.DOUBLE.equals(numberType)) {
+          // handle double value conversion
+          return (float) Double.longBitsToDouble(docValues.longValue());
+        }
+        // just take the long value
+        return docValues.longValue();
+      }
```
1/2 So I was trying to learn and understand better w.r.t. why the `intBitsToFloat` and `longBitsToDouble` are the conversions to use here and that led to https://github.com/apache/lucene-solr/blob/releases/lucene-solr/8.8.2/solr/core/src/java/org/apache/solr/search/SolrDocumentFetcher.java#L603-L644 and the realisation that `DATE` is a `numberType` possibility!

Would you have any thoughts on using a `switch` statement here with FLOAT/DOUBLE/INT/LONG all valid and to throw an exception for anything else ...

cpoerschke replied:
Somewhat strangely so perhaps the use of `DATE` fields "just works" for both the existing stored and the new docValues implementations. I've added a `TestFieldValueFeature.testThatDateValuesAreCorrectlyParsed` test to demonstrate that and so then no need to change the implementation here.

tomglk replied:
Hm... the date being a numeric type really is quite unintuitive.
I am not really fond of copying the switch-logic that you referenced, but sadly the method is private. :/

I will go and make a tea and then look into that.
Thank you for adding the test for the date field. The new clarity on what value is returned for the date type is very good.

tomglk replied:
Oh... Thanks github for not refreshing the page... :roll_eyes: 
After looking into that I saw the same behavior as you did and was just about to comment that. :D

cpoerschke replied:
> ... I am not really fond of copying the switch-logic that you referenced, but sadly the method is private. ...

Likewise, I would have preferred for there to be some existing reusable method and was surprised that there isn't one (or not an obviously findable one). The `SolrDocumentFetcher` one has the `sortableNumeric` flag (which we don't need here) and some special logic for `LatLonPointSpatialField` and `AbstractEnumField` scenarios (also not applicable here) ... otherwise it would have been a clear _"okay, let's factor out a method here and put it somewhere that both SolrDocumentFeature and FieldValueFeature can use it"_ kind of scenario. Oh well.

cpoerschke commented on solr/contrib/ltr/src/java/org/apache/solr/ltr/feature/FieldValueFeature.java:
```diff
@@ -83,18 +104,52 @@ public FeatureWeight createWeight(IndexSearcher searcher, boolean needsScores,
+        final FieldInfo fieldInfo = context.reader().getFieldInfos().fieldInfo(field);
+        final DocValuesType docValuesType = fieldInfo != null ? fieldInfo.getDocValuesType() : DocValuesType.NONE;
+
+        if (DocValuesType.NUMERIC.equals(docValuesType)) {
+          return new NumericDocValuesFieldValueFeatureScorer(this, context,
+                  DocIdSetIterator.all(DocIdSetIterator.NO_MORE_DOCS), schemaField.getType().getNumberType());
+        } else if (DocValuesType.SORTED.equals(docValuesType)) {
+          return new SortedDocValuesFieldValueFeatureScorer(this, context,
+                  DocIdSetIterator.all(DocIdSetIterator.NO_MORE_DOCS));
+        } else if (DocValuesType.NONE.equals(docValuesType)) {
+          // Using a fallback feature scorer because this segment has no documents with a doc value for the current field
+          return new DefaultValueFieldValueFeatureScorer(this, DocIdSetIterator.all(DocIdSetIterator.NO_MORE_DOCS));
+        }
```
2/2 ... combined with here diverting any unexpected `getNumberType()` enums onto the "is not supported" code path? Although having said that, I don't know what the existing `FieldValueFeatureScorer` would return for a stored `DATE` field i.e. perhaps back compat considerations apply, hmm.

tomglk commented:
Hi @cpoerschke! :)
> My conclusion was that for back compat reasons the docValues would need to match stored fields behaviour -- _I will add a commit with proposed changes shortly_ -- and that what looks like a "string" value in the implementation is for "boolean" values only really i.e. F/0 and T/1 and when number values are required a number field type would/should be used instead.

I just had the time to look through your changes.
Although I liked the possibility to also handle numbers that are present in a string field, I agree that one should use a numeric field for that. One shouldn't expect the code to handle such cases.
I think your changes made the functionalities that are supported much cleaner. And the consistent behavior between dv and non-dv fields also is a very important point that I admittedly paid not enough attention to. Thanks for keeping an eye on such things!

> _And advance apologies, the commit has multiple things in one commit though I've tried to at least describe in the commit message details what is in it._

No worries, I read you notes in the commit while looking through the code and had no problem with understanding your thoughts behind the changes.

cpoerschke commented:
> ... I just had the time to look through your changes. ...

Thanks @tomglk for taking a look!

I've added one more small polish commit to workaround us not yet having 'spotless' or other automatic code formatting for the `contrib/ltr` though #126 hopes to be a step in that direction.

Otherwise I think we're all done here, what do you think?

tomglk commented:
Thank you for the polishing and please excuse me for adding some spaces in these places. I tried to set up my IDE to mimic the correct formatting, but did not fully succeed.
#126 sounds like a good approach to get to a point where we don't have many formatting changes that blow up the change list of a PR. That would make everything easier.

> Otherwise I think we're all done here, what do you think?

I think so too. :)

tomglk commented:
Thank you for the great collaboration on this PR, @cpoerschke . I really did learn a lot thanks to your constructive remarks and had a lot of fun working on this. :)
